### ISSUE TITLE
exercise 5.3

### ISSUE THREAD

**Issue by michaelw123** [fpinscala/fpinscala#489]
takeWhile needs another case:
case Cons(h,t) if (!f(h())) => t() takeWhile f

**Comment by arobson73** [comment]
the solution does have another case
case _ => empty

**Comment by ekr387** [comment]
I think that @michaelw123 is right. For instance, this test fails:

```
Stream(1, 2, 3, 4, 5, 6).takeWhile(a => a % 2 == 0).toList shouldBe List(2, 4, 6)
```

As soon as the `1` is evaluated we get into the `case _ => empty` branch (because `1` is not even) and we're done. The test fails with:
```
Expected :List(2, 4, 6)
Actual   :List()
```

**Comment by thomhickey** [comment]
Yes, answer in the book doesn't pass my tests either. It doesn't handle the case where the predicate fails, so it returns an empty stream unless all elements in the stream pass the predicate.

**Comment by thomhickey** [comment]
takeWhile only takes from the start of the stream until the predicate fails.
